File: frontend/apps/ios/Pantopus/Core/Networking/Models/Notifications/NotificationsDTOs.swift
//
//  NotificationsDTOs.swift
//  Pantopus
//
//  Decoder shapes for `/api/notifications/*`.
//

import Foundation

public struct NotificationsListResponse: Decodable, Sendable {
    public let notifications: [NotificationDTO]
    public let unreadCount: Int?
    public let hasMore: Bool?
}

public struct NotificationDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let userId: String?
    public let type: String?
    public let title: String?
    public let body: String?
    public let icon: String?
    /// Backend-emitted deep link path, e.g. `/post/abc-123`,
    /// `/homes/h_1/dashboard`. DeepLinkRouter parses this.
    public let link: String?
    public let isRead: Bool?
    public let createdAt: String?
    public let context: String?

    enum CodingKeys: String, CodingKey {
        case id, type, title, body, icon, link, context
        case userId = "user_id"
        case isRead = "is_read"
        case createdAt = "created_at"
    }
}

public struct NotificationUnreadCountResponse: Decodable, Sendable {
    public let count: Int
}

public struct NotificationActionEcho: Decodable, Sendable {
    public let ok: Bool?
    public let count: Int?
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/notifications/NotificationsDtos.kt
package app.pantopus.android.data.api.models.notifications

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/** Single row from `/api/notifications`. */
@JsonClass(generateAdapter = true)
data class NotificationDto(
    val id: String,
    @Json(name = "user_id") val userId: String?,
    val type: String?,
    val title: String?,
    val body: String?,
    val icon: String?,
    /**
     * Backend-emitted deep link path, e.g. `/post/abc-123`,
     * `/homes/h_1/dashboard`. DeepLinkRouter parses this.
     */
    val link: String?,
    @Json(name = "is_read") val isRead: Boolean?,
    @Json(name = "created_at") val createdAt: String?,
    val context: String? = null,
)

/** `GET /api/notifications` envelope — route `backend/routes/notifications.js:84`. */
@JsonClass(generateAdapter = true)
data class NotificationsListResponse(
    val notifications: List<NotificationDto> = emptyList(),
    val unreadCount: Int? = null,
    val hasMore: Boolean? = null,
)

/** `GET /api/notifications/unread-count` envelope — route `backend/routes/notifications.js:160`. */
@JsonClass(generateAdapter = true)
data class NotificationUnreadCountResponse(
    val count: Int,
)

/** Echo of a write call (`/read` / `/read-all`). Both `ok` and `count` are optional on success. */
@JsonClass(generateAdapter = true)
data class NotificationActionEcho(
    val ok: Boolean? = null,
    val count: Int? = null,
)
